Step 1: Import relevant libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

Step 2 : Load Dataset

In [2]:
data = pd.read_csv('TMDB_movie_dataset_v11.csv')

Step 3: Overview of dataset

In [3]:
print(data.shape)

(994914, 23)


In [4]:
data.info(23)

<class 'pandas.DataFrame'>
RangeIndex: 994914 entries, 0 to 994913
Data columns (total 23 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   id                    994914 non-null  int64  
 1   title                 994902 non-null  str    
 2   vote_average          994914 non-null  float64
 3   vote_count            994914 non-null  int64  
 4   status                994914 non-null  str    
 5   release_date          882939 non-null  str    
 6   revenue               994914 non-null  int64  
 7   runtime               994914 non-null  int64  
 8   adult                 994914 non-null  bool   
 9   backdrop_path         282996 non-null  str    
 10  budget                994914 non-null  int64  
 11  homepage              107815 non-null  str    
 12  imdb_id               569928 non-null  str    
 13  original_language     994914 non-null  str    
 14  original_title        994902 non-null  str    
 15  overview   

In [5]:
data.head(10)

,id,title,vote_average,vote_count,status,release_date,revenue,runtime,adult,backdrop_path,...,original_language,original_title,overview,popularity,poster_path,tagline,genres,production_companies,production_countries,spoken_languages
0,27205,Inception,8.364,34495,Released,2010-07-15,825532764,148,False,/8ZTVqvKDQ8emSGUEMjsS4yHAwrp.jpg,...,en,Inception,"Cobb, a skilled thief who commits corporate es...",83.952,/oYuLEt3zVCKq57qu2F8dT7NIa6f.jpg,Your mind is the scene of the crime.,"Action, Science Fiction, Adventure","Legendary Pictures, Syncopy, Warner Bros. Pict...","United Kingdom, United States of America","English, French, Japanese, Swahili"
1,157336,Interstellar,8.417,32571,Released,2014-11-05,701729206,169,False,/pbrkL804c8yAv3zBZR4QPEafpAR.jpg,...,en,Interstellar,The adventures of a group of explorers who mak...,140.241,/gEU2QniE6E77NI6lCU6MxlNBvIx.jpg,Mankind was born on Earth. It was never meant ...,"Adventure, Drama, Science Fiction","Legendary Pictures, Syncopy, Lynda Obst Produc...","United Kingdom, United States of America",English
2,155,The Dark Knight,8.512,30619,Released,2008-07-16,1004558444,152,False,/nMKdUUepR0i5zn0y1T4CsSB5chy.jpg,...,en,The Dark Knight,Batman raises the stakes in his war on crime. ...,130.643,/qJ2tW6WMUDux911r6m7haRef0WH.jpg,Welcome to a world without rules.,"Drama, Action, Crime, Thriller","DC Comics, Legendary Pictures, Syncopy, Isobel...","United Kingdom, United States of America","English, Mandarin"
3,19995,Avatar,7.573,29815,Released,2009-12-15,2923706026,162,False,/vL5LR6WdxWPjLPFRLe133jXWsh5.jpg,...,en,Avatar,"In the 22nd century, a paraplegic Marine is di...",79.932,/kyeqWdyUXW608qlYkRqosgbbJyK.jpg,Enter the world of Pandora.,"Action, Adventure, Fantasy, Science Fiction","Dune Entertainment, Lightstorm Entertainment, ...","United States of America, United Kingdom","English, Spanish"
4,24428,The Avengers,7.710,29166,Released,2012-04-25,1518815515,143,False,/9BBTo63ANSmhC4e6r62OJFuK2GL.jpg,...,en,The Avengers,When an unexpected enemy emerges and threatens...,98.082,/RYMX2wcKCBAr24UyPD7xwmjaTn.jpg,Some assembly required.,"Science Fiction, Action, Adventure",Marvel Studios,United States of America,"English, Hindi, Russian"
5,293660,Deadpool,7.606,28894,Released,2016-02-09,783100000,108,False,/en971MEXui9diirXlogOrPKmsEn.jpg,...,en,Deadpool,The origin story of former Special Forces oper...,72.735,/zq8Cl3PNIDGU3iWNRoc5nEZ6pCe.jpg,Witness the beginning of a happy ending.,"Action, Adventure, Comedy","20th Century Fox, The Donners' Company, Genre ...",United States of America,English
6,299536,Avengers: Infinity War,8.255,27713,Released,2018-04-25,2052415039,149,False,/mDfJG3LC3Dqb67AZ52x3Z0jU0uB.jpg,...,en,Avengers: Infinity War,As the Avengers and their allies have continue...,154.340,/7WsyChQLEftFiDOVTGkv3hFpyyt.jpg,An entire universe. Once and for all.,"Adventure, Action, Science Fiction",Marvel Studios,United States of America,"English, Xhosa"
7,550,Fight Club,8.438,27238,Released,1999-10-15,100853753,139,False,/hZkgoQYus5vegHoetLkCJzb17zJ.jpg,...,en,Fight Club,A ticking-time-bomb insomniac and a slippery s...,69.498,/pB8BM7pdSp6B6Ih7QZ4DrQ3PmJK.jpg,Mischief. Mayhem. Soap.,Drama,"Regency Enterprises, Fox 2000 Pictures, Taurus...",United States of America,English
8,118340,Guardians of the Galaxy,7.906,26638,Released,2014-07-30,772776600,121,False,/uLtVbjvS1O7gXL8lUOwsFOH4man.jpg,...,en,Guardians of the Galaxy,"Light years from Earth, 26 years after being a...",33.255,/r7vmZjiyZw9rpJMQJdXpjgiCOk9.jpg,All heroes start somewhere.,"Action, Science Fiction, Adventure",Marvel Studios,United States of America,English
9,680,Pulp Fiction,8.488,25893,Released,1994-09-10,213900000,154,False,/suaEOtk1N1sgg2MTM7oZd2cfVp3.jpg,...,en,Pulp Fiction,"A burger-loving hit man, his philosophical par...",74.862,/d5iIlFn5s0ImszYzBPb8JPIfbXD.jpg,Just because you are a character doesn't mean ...,"Thriller, Crime","Miramax, A Band Apart, Jersey Films",United States of America,"English, Spanish, French"


Step 4: Filter dataset to a more manageable size

In [6]:
filtered = data[
    (data['status'] == 'Released') &
    (data['vote_count'] >= 50) &
    (data['overview'].notna()) &
    (data['genres'].notna())
].copy()

print(filtered.shape)

(27844, 23)


Step 5: Reset and select relevant columns

In [7]:
filtered = filtered.reset_index(drop=True)

movies = filtered[['id', 'title', 'genres', 'overview', 'tagline', 'vote_average', 'vote_count']].copy()
movies.head(10)

,id,title,genres,overview,tagline,vote_average,vote_count
0,27205,Inception,"Action, Science Fiction, Adventure","Cobb, a skilled thief who commits corporate es...",Your mind is the scene of the crime.,8.364,34495
1,157336,Interstellar,"Adventure, Drama, Science Fiction",The adventures of a group of explorers who mak...,Mankind was born on Earth. It was never meant ...,8.417,32571
2,155,The Dark Knight,"Drama, Action, Crime, Thriller",Batman raises the stakes in his war on crime. ...,Welcome to a world without rules.,8.512,30619
3,19995,Avatar,"Action, Adventure, Fantasy, Science Fiction","In the 22nd century, a paraplegic Marine is di...",Enter the world of Pandora.,7.573,29815
4,24428,The Avengers,"Science Fiction, Action, Adventure",When an unexpected enemy emerges and threatens...,Some assembly required.,7.710,29166
5,293660,Deadpool,"Action, Adventure, Comedy",The origin story of former Special Forces oper...,Witness the beginning of a happy ending.,7.606,28894
6,299536,Avengers: Infinity War,"Adventure, Action, Science Fiction",As the Avengers and their allies have continue...,An entire universe. Once and for all.,8.255,27713
7,550,Fight Club,Drama,A ticking-time-bomb insomniac and a slippery s...,Mischief. Mayhem. Soap.,8.438,27238
8,118340,Guardians of the Galaxy,"Action, Science Fiction, Adventure","Light years from Earth, 26 years after being a...",All heroes start somewhere.,7.906,26638
9,680,Pulp Fiction,"Thriller, Crime","A burger-loving hit man, his philosophical par...",Just because you are a character doesn't mean ...,8.488,25893


Step 6: Handle any missing values in the tagline, overview and genres columns

In [8]:
movies['tagline'] = movies['tagline'].fillna('')
movies['overview'] = movies['overview'].fillna('')
movies['genres'] = movies['genres'].fillna('')

movies.isnull().sum()

id              0
title           0
genres          0
overview        0
tagline         0
vote_average    0
vote_count      0
dtype: int64

Step 7: Build a combined content feature

In [9]:
movies['content'] = movies['genres'] + ' ' + movies['overview'] + ' ' + movies['tagline']
movies[['title', 'content']].head()

,title,content
0,Inception,"Action, Science Fiction, Adventure Cobb, a ski..."
1,Interstellar,"Adventure, Drama, Science Fiction The adventur..."
2,The Dark Knight,"Drama, Action, Crime, Thriller Batman raises t..."
3,Avatar,"Action, Adventure, Fantasy, Science Fiction In..."
4,The Avengers,"Science Fiction, Action, Adventure When an une..."


Step 8: Use td-idf to vectorize combined content

In [10]:
from sklearn.feature_extraction.text import TfidfVectorizer

vectorizer = TfidfVectorizer(stop_words='english', max_features=10000)
tfidf_matrix = vectorizer.fit_transform(movies['content'])

print(tfidf_matrix.shape)

(27844, 10000)


Step 9: Calculate cosine similarity based on previous vectors

In [11]:
from sklearn.metrics.pairwise import cosine_similarity

cosine_sim = cosine_similarity(tfidf_matrix, tfidf_matrix)

print(cosine_sim.shape)

(27844, 27844)


Step 10: Build recommender function

In [12]:
# Build a mapping from movie title to its row index
indices = pd.Series(movies.index, index=movies['title']).drop_duplicates()

def recommend(title, n=10):
    if title not in indices:
        return f"'{title}' not found in the dataset."
    
    idx = indices[title]
    
    # Get similarity scores for this movie against all others
    sim_scores = list(enumerate(cosine_sim[idx]))
    
    # Sort by similarity score, descending
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
    
    # Skip the first result (it's the movie itself, similarity = 1.0)
    sim_scores = sim_scores[1:n+1]
    
    movie_indices = [i[0] for i in sim_scores]
    
    return movies[['title', 'genres', 'vote_average']].iloc[movie_indices]

Step 11: Test with sample queries

In [13]:
recommend('Inception')

,title,genres,vote_average
26823,Major Grom,"Action, Adventure, Science Fiction, Drama",6.764
4902,Possessor,"Horror, Thriller, Science Fiction",6.364
404,Transformers: Revenge of the Fallen,"Science Fiction, Action, Adventure",6.200
15958,Altitude,"Action, Thriller",4.915
20933,Shadow Man,"Action, Adventure, Thriller",4.348
6383,Cypher,"Thriller, Science Fiction",6.654
10592,Renaissance,"Action, Animation, Science Fiction",6.543
6566,House,"Comedy, Horror",6.096
13554,Repeaters,"Drama, Action, Crime, Science Fiction",5.800
19963,The Monkey's Paw,"Horror, Thriller",4.638


In [14]:
recommend('The Dark Knight')

,title,genres,vote_average
6932,"Batman: The Long Halloween, Part Two","Animation, Mystery, Action, Crime",7.459
6084,"Batman: The Long Halloween, Part One","Animation, Mystery, Action, Crime",7.504
25,The Dark Knight Rises,"Action, Crime, Drama, Thriller",7.777
879,Batman Forever,"Action, Crime, Fantasy",5.409
487,Batman,"Fantasy, Action, Crime",7.200
2540,Batman: The Killing Joke,"Action, Animation, Crime, Drama",6.597
3091,"Batman: The Dark Knight Returns, Part 2","Science Fiction, Action, Animation, Mystery",7.930
366,The Batman,"Crime, Mystery, Thriller",7.705
8916,The Batman vs. Dracula,"Animation, Action, Fantasy, Mystery, Horror",6.974
26068,Batman Beyond,"Action, Animation, Science Fiction, TV Movie",6.929


In [15]:
recommend('Interstellar')

,title,genres,vote_average
4085,Space Pirate Captain Harlock,"Animation, Science Fiction",6.623
217,Prometheus,"Science Fiction, Adventure, Mystery",6.535
13205,The Beyond,Science Fiction,5.500
678,Gattaca,"Thriller, Science Fiction, Mystery, Romance",7.548
3964,The Day the Earth Stood Still,"Drama, Science Fiction, Thriller",7.472
26486,AE: Apocalypse Earth,"TV Movie, Science Fiction",3.972
17344,On the Silver Globe,"Science Fiction, Adventure, Fantasy, Drama",7.757
9609,Voices of a Distant Star,"Animation, Science Fiction, Romance",6.480
12416,3022,Science Fiction,5.402
5743,Breach,"Science Fiction, Action, Horror, Adventure",4.262


In [16]:
recommend('Iron Man')

,title,genres,vote_average
26,Iron Man 3,"Action, Adventure, Science Fiction",6.928
40,Iron Man 2,"Adventure, Action, Science Fiction",6.835
3283,Clown,"Horror, Thriller",5.747
11296,The Invincible Iron Man,"Action, Adventure, Animation, Fantasy, Science...",6.124
23,Avengers: Age of Ultron,"Action, Adventure, Science Fiction",7.276
17987,Inhumans: The First Chapter,"Science Fiction, Action, Adventure",5.297
20833,Fatal Fury: The Motion Picture,"Action, Animation, Fantasy",7.638
8691,Shut In,"Thriller, Horror, Drama",6.935
6158,Igor,"Animation, Comedy, Family, Fantasy, Horror",5.958
25435,Sharktopus vs. Whalewolf,"Action, Science Fiction, Adventure, TV Movie, ...",3.638


In [17]:
recommend('Ice Age')

,title,genres,vote_average
431,Ice Age: Dawn of the Dinosaurs,"Animation, Comedy, Family, Adventure",6.737
334,Ice Age: The Meltdown,"Animation, Family, Comedy, Adventure",6.688
5736,Ice Age: A Mammoth Christmas,"Animation, Family, TV Movie",6.413
496,Ice Age: Continental Drift,"Animation, Comedy, Adventure, Family",6.360
26398,Eegah,Horror,2.620
1017,Ice Age: Collision Course,"Adventure, Animation, Family, Comedy, Science ...",6.073
25585,Ghosthunters: On Icy Trails,"Comedy, Family, Fantasy",5.667
18602,Ice Age: Surviving Sid,"Animation, Comedy, TV Movie",6.094
24151,The Deadly Mantis,"Horror, Science Fiction",4.873
4758,Two Brothers,"Adventure, Drama, Family",7.134


Qualitative Evaluation

Testing the recommender on several sample movies showed a mix of strong and weaker results. Franchise-based queries (e.g. "The Dark Knight", "Ice Age") produced excellent recommendations, almost entirely surfacing other films in the same franchise. This is likely because shared vocabulary (character names, franchise terms) in the overview/tagline text reinforced genre similarity.

Standalone films with more unique thematic content (e.g. "Inception") produced recommendations that matched on genre but didn't fully capture the film's distinctive narrative concepts, since those concepts aren't always explicitly stated in the overview text. Occasional unrelated matches also appeared (e.g. a horror film recommended alongside Iron Man), reflecting a known limitation of content-based filtering: it can only detect similarity based on the literal text provided, not deeper semantic or thematic meaning.

Overall, the recommender performs well for franchise and genre-consistent queries, with room for improvement on more nuanced, standalone films.

Step 12: Flask API Deployment

To demonstrate the recommender as a deployable service rather than just a notebook, the model was wrapped in a Flask API with a single /recommend endpoint. It accepts a title query parameter and optional n for number of results and returns JSON-formatted recommendations.

Example request: http://127.0.0.1:5000/recommend?title=The%20Dark%20Knight&n=5

Testing confirmed the API correctly reproduces the notebook's recommendations, successfully returning results for both "Inception" and "The Dark Knight" queries.

Conclusion

This project built a content-based movie recommendation system using the Full TMDB Movies Dataset 2024. Starting from nearly 1 million movie records, the dataset was filtered down to 27,844 released movies with sufficient vote counts and complete overview/genre metadata. This was a necessary step, since computing pairwise similarity across the full dataset might have been difficult to do.

Genre, overview, and tagline text were combined into a single content field per movie, then converted into numeric features using TF-IDF vectorization. Cosine similarity was computed across all movies, and a recommendation function was built to return the top N most similar titles for any given movie.

Qualitative evaluation: Across several sample queries showed strong performance on franchise-based films like "The Dark Knight" and "Ice Age" both returned highly relevant sequels and spin-offs, while standalone films with more distinctive narrative concepts like "Inception" for example produced recommendations that matched on genre but didn't always capture deeper thematic similarity. This is an expected limitation of content-based filtering using only genre, overview, and tagline text, since the model can only detect similarity present in the literal text provided, without access to keywords or cast/crew data that could strengthen thematic matching.

Flask extension: the recommender was packaged as a Flask API with a /recommend endpoint, accepting a movie title and returning JSON-formatted recommendations. This demonstrates the model functioning as a deployable service rather than only within a notebook environment, successfully reproducing the same recommendations available in the core analysis.

Limitations and future improvements:
1. Incorporating cast, crew, or keyword data which was not available in this dataset could improve thematic matching for standalone films
2. Using n-grams or weighting overview text more heavily could help capture multi-word concepts currently diluted by single-word TF-IDF features
3. A hybrid approach combining content-based filtering with collaborative filtering using user rating data could address cases where content alone isn't sufficient to capture what makes two movies genuinely similar to viewers
4. The Flask API is currently a local development server only; production deployment would require a proper WSGI server and likely a pre-computed, saved similarity matrix rather than recalculating it on every startup

Overall, this project demonstrates a functional, well-evaluated content-based recommendation system, with an honest assessment of where it performs strongly  and where it performed weakly, along with a working deployment as a proof of concept.